### PR TITLE
Implements `IO` monad.

### PR DESCRIPTION
**Pull request by drupol**
This PR implements the `IO` monad.

Keep in mind that I'm doing this to learn monads, it might contains inconsistencies - please bear with me.

### CODE REVIEW

File: src/IO.php [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+use Marcosh\LamPHPda\Brand\IOBrand;
+
+/**
+ * @template A
+ * @implements DefaultMonad<IOBrand, A>
+ */
+final class IO implements DefaultMonad
+{
+    /**
+     * @var callable
```
Comment (marcosh) [inline src/IO.php]: you could be more precise using Psalm. This should be a `callable(): A`

File: src/IO.php [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    private function __construct(callable $f)
+    {
+        $this->unsafe = $f;
+    }
+
+    /**
+     * @template B
+     * @param B $x
+     * @return IO<B>
```
Comment (marcosh) [inline src/IO.php:41]: shouldn't this be `pure`?

File: src/IO.php [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        return $b;
+    }
+
+    /**
+     * @template B
+     * @param Monad<IOBrand> $monad
+     * @param callable(A): HK1<IOBrand, B> $f
+     * @return IO<B>
+     */
+    public function ibind(Monad $monad, callable $f): IO
```
Comment (marcosh) [inline src/IO.php:67]: please keep the order we have in the other monads, i.e. first `map`, then `apply`, `pure` and `bind`

File: src/IO.php (hunk 2 above)
Comment (marcosh) [reply src/IO.php:41]: in general, let's try to have purity annotations everywhere

File: src/IO.php [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+     * @return IO<B>
+     *
+     * @psalm-suppress LessSpecificImplementedReturnType
+     */
+    public static function pure($a): self
+    {
+        return self::ipure(new IOApplicative(), $a);
+    }
+
+        /**
```
Comment (marcosh) [inline src/IO.php:108]: oops, wrong indentation

File: src/IO.php [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+     *
+     * @psalm-suppress LessSpecificImplementedReturnType
+     */
+    public function map(callable $f): IO
+    {
+        return $this->imap(new IOFunctor(), $f);
+    }
+
+    /**
+     * @todo Should we have a parameter here?
```
Comment (marcosh) [inline src/IO.php:165]: no, no parameters are needed here, this amounts to invoking a function with no arguments

File: src/IO.php [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+     *
```
Comment (marcosh) [inline src/IO.php:167]: return type should be `A`

File: src/IO.php [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    {
+        $this->unsafe = $f;
+    }
+
+    /**
+     * @template B
+     * @param B $x
+     * @return IO<B>
+     */
+    public static function of($x): IO
```
Comment (marcosh) [inline src/IO.php:43]: instead of having `of`, why don't we use directly `pure`?

File: src/Instances/IO/IOMonad.php [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+    public function bind(HK1 $a, callable $f): IO
+    {
+        $ioA = IO::fromBrand($a);
+
+        $ioAEval = $ioA->eval();
+
+        /** @var IO<B> $ioAEvalWithF */
+        $ioAEvalWithF = $f($ioAEval);
+
+        $ioAEvalWithFEval = $ioAEvalWithF->eval();
```
Comment (marcosh) [inline src/Instances/IO/IOMonad.php:85]: doing `->eval()` and then `IO::of` should be a no-op, should be enough doing `return IO::fromBrand($f($ioA->eval()))`

File: src/IO.php (hunk 1 above)
Comment (drupol) [reply src/IO.php]: Done.

File: src/IO.php (hunk 3 above)
Comment (drupol) [reply src/IO.php]: Done.

File: src/IO.php (hunk 4 above)
Comment (drupol) [reply src/IO.php]: Done.

File: src/IO.php (hunk 5 above)
Comment (drupol) [reply src/IO.php]: Done.

File: src/IO.php (hunk 6 above)
Comment (drupol) [reply src/IO.php:167]: Done.

File: src/Instances/IO/IOMonad.php (hunk 8 above)
Comment (drupol) [reply src/Instances/IO/IOMonad.php]: Done.

**Comment by matyo91** [comment]
For my part, I did consider array of callable in the constructor as seen here : https://github.com/darkwood-fr/railway-fbp/blob/1.x/packages/php/src/Rail/Rail.php
I'm interested to see when it's merged. As here there are things that are better done with PSalm typing analyze. I may consider to integrate it for `Rail`. The main difference is I prefer call it `pipe` and not `bind` witch is more a monad approach. Don't know what to decide for now.

**Comment by marcosh** [comment]
> For my part, I did consider array of callable in the constructor as seen here : https://github.com/darkwood-fr/railway-fbp/blob/1.x/packages/php/src/Rail/Rail.php I'm interested to see when it's merged. As here there are things that are better done with PSalm typing analyze. I may consider to integrate it for `Rail`. The main difference is I prefer call it `pipe` and not `bind` witch is more a monad approach. Don't know what to decide for now.

Thanks @matyo91 for the feedback! What do you see as the benefit of having an array of callables instead of just one, considering that they can be composed?

This implementation is inspired by https://gcanti.github.io/fp-ts/modules/IO.ts.html

I prefer to use Haskell names, because they are the ones I'm most familiar with, and also well known in the literature.

Moreover, I see pipe more as https://hackage.haskell.org/package/base-4.16.0.0/docs/Control-Monad.html#v:-62--61--62- than `bind`

**Comment by matyo91** [comment]
> Thanks @matyo91 for the feedback! What do you see as the benefit of having an array of callables instead of just one, considering that they can be composed?

In my Rail approach, having an `array` of `jobs` (in my terms) was more to make the computation more generic. But i'm Ok to say that it's equivalent to have only one `job` instead of multiple `jobs`. The only difference is when you look inside `Rail::nextIpJob` there is asynchronous management by using `coroutine` so it's likely implementing Promise::all for `Rail`. I could even think about having an option in the constructor to make Promise::race (it would mean that first job done is calling the next pipe).
For now, it's completely theorical, as I get zero practical implementation and usage. I was given `array of jobs` just because this could be seamly directly integrated without many overhead.

> This implementation is inspired by https://gcanti.github.io/fp-ts/modules/IO.ts.html

Good to know if I want to transpile Railway-FBP to Typescript.

> I prefer to use Haskell names, because they are the ones I'm most familiar with, and also well known in the literature.
> 
> Moreover, I see pipe more as https://hackage.haskell.org/package/base-4.16.0.0/docs/Control-Monad.html#v:-62--61--62- than `bind`

Yes the fact in Haskell `>=>` is an element of the language and not a `word` so this could lead to interpretation. Yes I found out that it's different from `>>=` notation.

I so a smart `do` implementation in PHP with yield syntax here : https://github.com/whsv26/functional/blob/14957faba58044deab7ab23fd7f00466e98445e9/doc/Monads.md

**Comment by marcosh** [comment]
> In my Rail approach, having an `array` of `jobs` (in my terms) was more to make the computation more generic. But i'm Ok to say that it's equivalent to have only one `job` instead of multiple `jobs`. The only difference is when you look inside `Rail::nextIpJob` there is asynchronous management by using `coroutine` so it's likely implementing Promise::all for `Rail`. I could even think about having an option in the constructor to make Promise::race (it would mean that first job done is calling the next pipe). For now, it's completely theorical, as I get zero practical implementation and usage. I was given `array of jobs` just because this could be seamly directly integrated without many overhead.

using coroutines seems a bit out of scope for what I'm doing here. I haven't really thought about async computations, but if and when I'll do it, I think I'll draw my inspiration from https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html

> I so a smart `do` implementation in PHP with yield syntax here : https://github.com/whsv26/functional/blob/14957faba58044deab7ab23fd7f00466e98445e9/doc/Monads.md

That's really cute! I had my shot at mimicking do notation in http://marcosh.github.io/post/2019/04/10/applicative-monadic-validation-php.html

**Comment by matyo91** [comment]
> using coroutines seems a bit out of scope for what I'm doing here. I haven't really thought about async computations, but if and when I'll do it, I think I'll draw my inspiration from https://hackage.haskell.org/package/async-2.2.4/docs/Control-Concurrent-Async.html

Interesting. I'm looking forward to see how it will be implemented. In php, this is possible to do fork but you need use PHP extension api. For what I read on the article, I never write haskell in my life but seem's that this github repository `lamphpda` do take good translation from Haskell to PHP. In my case I did choose coroutine and factually this is not forking process. So there are some execution differences thread vs coroutine, but the `asynchronous modad`  part should feel the same. I think `lamphpda` is lower level than me in `RailwayFBP`. In my part I got also the `Information Packet` part that is not shown in pure monadic way. That's why i should have interest to see if my `Rail`
 could in the end extends or be an element of `withAsync` that you could implement a futur time.

> That's really cute! I had my shot at mimicking do notation in http://marcosh.github.io/post/2019/04/10/applicative-monadic-validation-php.html

Honesty I do prefer the `do` notation with `yield` syntax and generator. In your case you define two : `mdo` and `sdo` !

**Comment by marcosh** [comment]
@drupol I completely forgot about this PR and I implemented `IO` myself, sorry about this.

I'm going to close this PR

**Pull request closed without merging**